OpenRA.Mods.Common/Traits/Conditions/ExternalCondition.cs

```diff
@@ -0,0 +1,142 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Allows a condition to be granted from an external source (Lua, warheads, etc).")]
+	public class ExternalConditionInfo : ITraitInfo, Requires<ConditionManagerInfo>
+	{
+		[GrantedConditionReference]
+		[FieldLoader.Require]
+		public readonly string Condition = null;
+
+		[Desc("If > 0, restrict the number of times that this condition can be granted by a single source.")]
+		public readonly int SourceCap = 0;
+
+		[Desc("If > 0, restrict the number of times that this condition can be granted by any source.")]
+		public readonly int TotalCap = 0;
+
+		public object Create(ActorInitializer init) { return new ExternalCondition(init.Self, this); }
+	}
+
+	public class ExternalCondition
+	{
+		class TimedToken
+		{
+			public int Token;
+			public int Expires;
+		}
+
+		public readonly ExternalConditionInfo Info;
+		readonly ConditionManager conditionManager;
+		readonly Dictionary<object, HashSet<int>> permanentTokens = new Dictionary<object, HashSet<int>>();
+		readonly Dictionary<object, HashSet<TimedToken>> timedTokens = new Dictionary<object, HashSet<TimedToken>>();
+
+		public ExternalCondition(Actor self, ExternalConditionInfo info)
+		{
+			Info = info;
+			conditionManager = self.Trait<ConditionManager>();
+		}
+
+		public bool CanGrantCondition(Actor self, object source, int duration = 0)
+		{
+			if (conditionManager == null || source == null)
+				return false;
+
+			// Timed tokens do not count towards the source cap: the condition with the shortest
+			// remaining duration can always be revoked to make room.
+			if (Info.SourceCap > 0 && permanentTokens.GetOrAdd(source).Count >= Info.SourceCap)
```

Comment:
Ah, permanentTokens. Nevermind me then. The value of duration went unused however.